src/Forms/SearchableDropdownTrait.php

```diff
@@ -0,0 +1,594 @@
+
+use ArrayAccess;
+use Error;
+use InvalidArgumentException;
+use LogicException;
+use SilverStripe\Control\Controller;
+use SilverStripe\Control\HTTPRequest;
+use SilverStripe\Control\HTTPResponse;
+use SilverStripe\ORM\ArrayList;
+use SilverStripe\ORM\DataList;
+use SilverStripe\ORM\DataObject;
+use SilverStripe\ORM\DataObjectInterface;
+use SilverStripe\ORM\Relation;
+use SilverStripe\ORM\SS_List;
+use SilverStripe\ORM\FieldType\DBHTMLText;
+use SilverStripe\View\ArrayData;
+use SilverStripe\ORM\Search\SearchContext;
+use SilverStripe\Security\SecurityToken;
+use Exception;
+
+trait SearchableDropdownTrait
+{
+    private static array $allowed_actions = [
+        'search',
+    ];
+
+    private bool $isClearable = false;
+
+    private bool $isLazyLoaded = false;
+
+    private bool $isMultiple = false;
+
+    private bool $isSearchable = true;
+
+    private int $lazyLoadLimit = 10;
+
+    private string $placeholder = '';
+
+    private ?SearchContext $searchContext = null;
+
+    private bool $useDynamicPlaceholder = true;
+
+    private bool $useSearchContext = false;
+
+    private DataList $sourceList;
+
+    private string $labelField = 'Title';
+
+    /**
+     * @param string $name
+     * @param string $title
+     * @param DataList|ArrayList|ArrayAccess|array $source
+     * @param mixed $value
+     * @param string $labelField
+     */
+    public function __construct($name, $title = '', $source = [], $value = null, $labelField = 'Title')
+    {
+        // Note that an ancestor contstructor will call $this->setSource() which will call
+        // setSource() in this class which throw an exception if $source is not a DataList
+        parent::__construct($name, $title, $source, $value);
+        $this->setLabelField($labelField);
+        $this->addExtraClass('ss-searchable-dropdown-field');
+        $this->init();
+    }
+
+    /**
+     * Returns a JSON string of options for lazy loading.
+     */
+    public function search(HTTPRequest $request): HTTPResponse
+    {
+        $response = HTTPResponse::create();
+        $response->addHeader('Content-Type', 'application/json');
+        if (!SecurityToken::singleton()->checkRequest($request)) {
+            $response->setStatusCode(400);
+            $response->setBody(json_encode(['message' => 'Invalid CSRF token']));
+            return $response;
+        }
+        $term = $request->getVar('term') ?? '';
+        $options = $this->getOptionsForSearchRequest($term);
+        $response->setBody(json_encode($options));
+        return $response;
+    }
+
+    /**
+     * Get whether the currently selected value(s) can be cleared
+     */
+    public function getIsClearable(): bool
+    {
+        return $this->isClearable;
+    }
+
+    /**
+     * Set whether the currently selected value(s) can be cleared
+     */
+    public function setIsClearable(bool $isClearable): static
+    {
+        $this->isClearable = $isClearable;
+        return $this;
+    }
```

Comment:
Since we're using react-select it's really whether or no to pass the 'clearable' prop which controls if a big 'X' appears on the right hand side - https://react-select.com/home

There may be some user cases where it's no desirable so may as well make it easy to toggle it off

Reply:
Hmm. Seems like unnecessary complexity to me but since it's already in there I'll allow it - assuming there's a behat test for it in admin.